tools/git-utils.sh

```diff
@@ -0,0 +1,108 @@
+# doesn't do anything on its own.  must be sourced.
+
+# The script which sources this script must set the following variables:
+#   GITHUB_REPO = the relative repo name of the submodule on github
+#   SUBDIR = the location in the working tree where the submodule goes
+# We also expect `set -eu`.
+[ -n "${GITHUB_REPO}" ]
+[ -n "${SUBDIR}" ]
+
+GITHUB_BASE="${GITHUB_BASE:-cockpit-project/cockpit}"
+GITHUB_REPOSITORY="${GITHUB_BASE%/*}/${GITHUB_REPO}"
+HTTPS_REMOTE="https://github.com/${GITHUB_REPOSITORY}"
+
+CACHE_DIR="${XDG_CACHE_HOME-${HOME}/.cache}/cockpit-dev/${GITHUB_REPOSITORY}.git"
+
+if [ "${V-}" = 0 ]; then
+    message() { printf "  %-8s %s\n" "$1" "$2"; }
+    quiet='--quiet'
+else
+    message() { :; }
+    quiet=''
+fi
+
+# runs a git command on the cache dir
+git_cache() {
+    git --git-dir "${CACHE_DIR}" "$@"
+}
+
+# reads the named gitlink from the current state of the index
+# returns (ie: prints) a 40-character commit ID
+get_index_gitlink() {
+    if ! git ls-files -s "$1" | egrep -o '\<[[:xdigit:]]{40}\>'; then
+        echo "*** couldn't read gitlink for file $1 from the index" >&2
+        exit 1
+    fi
+}
+
+init_cache() {
+    if [ ! -d "${CACHE_DIR}" ]; then
+        message INIT "${CACHE_DIR}"
+        mkdir -p "${CACHE_DIR}"
+        git init --bare --template='' ${quiet} "${CACHE_DIR}"
+        git_cache remote add origin "${HTTPS_REMOTE}"
+    fi
+}
+
+# This checks if the given argument "$1" (already) exists in the repository
+check_ref() {
+    git_cache rev-parse --verify --quiet "$1" >/dev/null 2>&1
+}
+
+# Fetch a specific commit ID into the cache
+# Either we have this commit available locally (in which case this function
+# does nothing), or we need to fetch it.  There's no chance that the object
+# changed on the server, because we define it by its checksum.
+fetch_sha_to_cache() {
+    sha="$1"
+
+    # No "offline mode" here: we either have the commit, or we don't
+    if ! check_ref "${sha}"; then
+        init_cache
+        message FETCH "${SUBDIR}  [ref: ${sha}]"
+        git_cache fetch --no-tags ${quiet} origin "${sha}"
+        # tag it to keep it from being GC'd.
+        git_cache tag "sha-${sha}" "${sha}"
```

Comment:
Leftover from  previous conversation: It's not clear to me at all why we need this tag. If git wants to gc the local cache repo from time to time, that's actually desirable -- I see no point in actively preventing that. In fact, we should probably do something about not growing the local caches indefinitely.

Naturally on the github.com repo we want (and have) tags, and do garbage collection in a more controlled fashion. But AFAICS these local tags never get pushed to origin, they already exist there at this point?

Reply:
I want to add an explicit prune mechanism at a later date.  I'm too worried about git dropping commits between commands, or if a separate parallel command is running for another workdir and it accidentally triggers a GC, for example...